File: henn-faiss/experiments/example_henn.py
#!/usr/bin/env python3
"""
Example usage of the HENN (Hierarchical EpsNet Navigation Graph) index in Python.

This demonstrates how to use the new HENN index through the Faiss Python interface.
HENN is similar to HNSW but uses EPS-Net sampling for layer construction.
"""

import numpy as np
import faiss
import time

def generate_random_data(n, d):
    """Generate random normalized vectors."""
    np.random.seed(1234)
    data = np.random.random((n, d)).astype('float32')
    # Normalize for better performance
    faiss.normalize_L2(data)
    return data

def test_henn_flat():
    """Test basic HENN flat index functionality."""
    print("=== Testing IndexHENNFlat ===")
    
    d = 64          # dimension
    n = 10000       # database size  
    nq = 100        # number of queries
    k = 10          # number of nearest neighbors
    M = 16          # max connections per node
    
    print(f"Dataset: {n} vectors of dimension {d}")
    
    # Generate data
    xb = generate_random_data(n, d)
    xq = generate_random_data(nq, d)
    
    # Create HENN index
    index = faiss.IndexHENNFlat(d, M)
    
    # Configure HENN parameters
    index.set_epsnet_strategy("random")      # or "budget_aware"
    index.set_pgraph_algorithm("nsw")        # or "knn" 
    index.set_exp_decay(2)                   # layer size reduction factor
    index.henn.ef_construction = 200         # construction parameter
    index.henn.efSearch = 32                 # search parameter
    
    print(f"Index configuration:")
    print(f"  EPS-Net strategy: {index.epsnet_strategy}")  
    print(f"  Graph algorithm: {index.pgraph_algorithm}")
    print(f"  Exponential decay: {index.exp_decay}")
    print(f"  ef_construction: {index.henn.ef_construction}")
    print(f"  efSearch: {index.henn.efSearch}")
    
    # Add vectors to index
    print(f"Adding {n} vectors...")
    start_time = time.time()
    index.add(xb)
    build_time = time.time() - start_time
    print(f"Build time: {build_time:.3f} seconds")
    
    # Print index statistics
    print(f"Index built with {len(index.henn.layers)} layers")
    print(f"Entry point: {index.henn.entry_point}")
    print(f"Max level: {index.henn.max_level}")
    
    # Search
    print(f"\nSearching for {k} nearest neighbors...")
    start_time = time.time()
    D, I = index.search(xq, k)
    search_time = time.time() - start_time
    
    print(f"Search time: {search_time:.3f} seconds")
    print(f"Average time per query: {search_time/nq*1000:.2f} ms")
    
    # Print some results
    print(f"\nFirst query results:")
    for i in range(min(5, k)):
        print(f"  {i+1}: ID={I[0,i]}, distance={D[0,i]:.6f}")
    
    return index, xb, xq

def compare_with_hnsw():
    """Compare HENN with HNSW performance."""
    print("\n=== Comparing HENN vs HNSW ===")
    
    d = 128
    n = 20000
    nq = 100 
    k = 10
    M = 16
    
    # Generate data
    xb = generate_random_data(n, d)
    xq = generate_random_data(nq, d)
    
    # Test HNSW
    print("Testing HNSW...")
    index_hnsw = faiss.IndexHNSWFlat(d, M)
    index_hnsw.hnsw.efConstruction = 200
    index_hnsw.hnsw.efSearch = 32
    
    start_time = time.time()
    index_hnsw.add(xb)
    hnsw_build_time = time.time() - start_time
    
    start_time = time.time() 
    D_hnsw, I_hnsw = index_hnsw.search(xq, k)
    hnsw_search_time = time.time() - start_time
    
    # Test HENN
    print("Testing HENN...")
    index_henn = faiss.IndexHENNFlat(d, M)
    index_henn.set_epsnet_strategy("budget_aware")
    index_henn.set_pgraph_algorithm("nsw") 
    index_henn.henn.ef_construction = 200
    index_henn.henn.efSearch = 32
    
    start_time = time.time()
    index_henn.add(xb)
    henn_build_time = time.time() - start_time
    
    start_time = time.time()
    D_henn, I_henn = index_henn.search(xq, k)
    henn_search_time = time.time() - start_time
    
    # Compare results
    print(f"\nPerformance comparison:")
    print(f"HNSW - Build: {hnsw_build_time:.3f}s, Search: {hnsw_search_time:.3f}s")
    print(f"HENN - Build: {henn_build_time:.3f}s, Search: {henn_search_time:.3f}s")
    
    # Calculate recall (approximate)
    recall = 0
    for i in range(nq):
        intersect = len(set(I_henn[i]) & set(I_hnsw[i]))
        recall += intersect / k
    recall /= nq
    
    print(f"Approximate recall between HENN and HNSW: {recall:.3f}")

def test_different_configurations():
    """Test different HENN configurations."""
    print("\n=== Testing Different HENN Configurations ===")
    
    d = 64
    n = 5000
    nq = 50
    k = 10 
    M = 16
    
    xb = generate_random_data(n, d)
    xq = generate_random_data(nq, d)
    
    configs = [
        ("random", "nsw"),
        ("random", "knn"), 
        ("budget_aware", "nsw"),
        ("budget_aware", "knn")
    ]
    
    results = []
    
    for epsnet_strategy, pgraph_algorithm in configs:
        print(f"\nTesting: EPS-Net={epsnet_strategy}, Graph={pgraph_algorithm}")
        
        index = faiss.IndexHENNFlat(d, M)
        index.set_epsnet_strategy(epsnet_strategy)
        index.set_pgraph_algorithm(pgraph_algorithm)
        index.henn.ef_construction = 100
        index.henn.efSearch = 20
        
        # Build
        start_time = time.time()
        index.add(xb)
        build_time = time.time() - start_time
        
        # Search
        start_time = time.time()
        D, I = index.search(xq, k)
        search_time = time.time() - start_time
        
        print(f"  Build time: {build_time:.3f}s")
        print(f"  Search time: {search_time:.3f}s")
        print(f"  Layers: {len(index.henn.layers)}")
        
        results.append({
            'config': (epsnet_strategy, pgraph_algorithm),
            'build_time': build_time,
            'search_time': search_time, 
            'layers': len(index.henn.layers)
        })
    
    # Summary
    print(f"\nConfiguration Summary:")
    for result in results:
        config_str = f"{result['config'][0]}-{result['config'][1]}"
        print(f"  {config_str:20} | Build: {result['build_time']:.3f}s | "
              f"Search: {result['search_time']:.3f}s | Layers: {result['layers']}")

def test_henn_pq():
    """Test HENN with Product Quantization."""
    print("\n=== Testing IndexHENNPQ ===")
    
    d = 128
    n = 10000
    nq = 100
    k = 10
    M = 16
    pq_m = 8       # number of PQ segments  
    pq_nbits = 8   # bits per PQ code
    
    xb = generate_random_data(n, d)
    xq = generate_random_data(nq, d)
    
    # Create HENN-PQ index
    index = faiss.IndexHENNPQ(d, pq_m, M, pq_nbits)
    index.set_epsnet_strategy("random")
    index.set_pgraph_algorithm("nsw")
    index.henn.ef_construction = 200
    
    # Train the PQ quantizer
    print("Training PQ quantizer...")
    xt = generate_random_data(10000, d)  # training data
    start_time = time.time()
    index.train(xt)
    train_time = time.time() - start_time
    print(f"Training time: {train_time:.3f}s")
    
    # Add data
    print("Adding data...")
    start_time = time.time()
    index.add(xb) 
    build_time = time.time() - start_time
    
    # Search
    start_time = time.time()
    D, I = index.search(xq, k)
    search_time = time.time() - start_time
    
    print(f"HENN-PQ Results:")
    print(f"  Build time: {build_time:.3f}s")
    print(f"  Search time: {search_time:.3f}s")
    print(f"  Index size: ~{(n * pq_m * pq_nbits) / 8 / 1024:.1f} KB (compressed)")

def main():
    """Main function to run all tests."""
    print("HENN Index Python Examples")
    print("=" * 50)
    print("HENN: Hierarchical EpsNet Navigation Graph")
    print("Similar to HNSW but with EPS-Net sampling for layer construction")
    print()
    
    try:
        # Basic functionality test
        test_henn_flat()
        
        # Compare with HNSW
        compare_with_hnsw()
        
        # Different configurations
        test_different_configurations()
        
        # PQ variant
        test_henn_pq()
        
        print("\n" + "=" * 50)
        print("All HENN tests completed successfully!")
        print("\nHENN Index Features:")
        print("- Hierarchical structure with EPS-Net sampling")
        print("- Choice of 'random' or 'budget_aware' sampling")
        print("- Choice of 'nsw' or 'knn' proximity graphs")  
        print("- Configurable layer size reduction (exp_decay)")
        print("- Compatible with PQ and SQ compression")
        print("- Similar search performance to HNSW")
        
    except Exception as e:
        print(f"Error during testing: {e}")
        import traceback
        traceback.print_exc()

if __name__ == "__main__":
    main()

File: henn-faiss/experiments/example_henn.cpp
/*
 * Simple example program demonstrating how to use the HENN index in Faiss.
 * 
 * This example shows:
 * 1. Creating a HENN index with different configurations
 * 2. Adding vectors to the index
 * 3. Searching for nearest neighbors
 * 4. Configuring EPS-Net sampling and proximity graph algorithms
 */

#include <iostream>
#include <vector>
#include <random>
#include <chrono>

#include <faiss/IndexHENN.h>
#include <faiss/index_io.h>
#include <faiss/utils/distances.h>

void generate_random_data(std::vector<float>& data, int n, int d) {
    std::random_device rd;
    std::mt19937 gen(rd());
    std::normal_distribution<float> dis(0.0, 1.0);
    
    data.resize(n * d);
    for (int i = 0; i < n * d; i++) {
        data[i] = dis(gen);
    }
}

void test_henn_flat() {
    std::cout << "=== Testing IndexHENNFlat ===" << std::endl;
    
    int d = 64;    // dimension
    int n = 10000; // number of vectors
    int M = 16;    // max connections per node
    
    // Generate random data
    std::vector<float> database(n * d);
    generate_random_data(database, n, d);
    
    // Create HENN index
    faiss::IndexHENNFlat index(d, M, faiss::METRIC_L2);
    
    // Configure HENN parameters
    index.set_epsnet_strategy("random");      // or "budget_aware"
    index.set_pgraph_algorithm("nsw");        // or "knn"
    index.set_exp_decay(2);                   // exponential decay factor
    index.henn.ef_construction = 200;         // construction parameter
    index.henn.efSearch = 16;                 // default search parameter
    
    std::cout << "Adding " << n << " vectors to index..." << std::endl;
    auto start = std::chrono::high_resolution_clock::now();
    
    index.add(n, database.data());
    
    auto end = std::chrono::high_resolution_clock::now();
    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    std::cout << "Index construction took: " << duration.count() << " ms" << std::endl;
    
    // Print index statistics
    index.henn.print_stats();
    
    // Generate query vectors
    int nq = 100;
    std::vector<float> queries(nq * d);
    generate_random_data(queries, nq, d);
    
    // Search parameters
    int k = 10;
    std::vector<float> distances(nq * k);
    std::vector<faiss::idx_t> labels(nq * k);
    
    // Create search parameters
    faiss::SearchParametersHENN search_params;
    search_params.ef = 50;  // search parameter
    
    std::cout << "Searching for " << k << " nearest neighbors for " << nq << " queries..." << std::endl;
    start = std::chrono::high_resolution_clock::now();
    
    index.search(nq, queries.data(), k, distances.data(), labels.data(), &search_params);
    
    end = std::chrono::high_resolution_clock::now();
    duration = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
    std::cout << "Search took: " << duration.count() << " μs" << std::endl;
    std::cout << "Average time per query: " << duration.count() / nq << " μs" << std::endl;
    
    // Print some results
    std::cout << "First query results:" << std::endl;
    for (int i = 0; i < k; i++) {
        std::cout << "  " << i << ": label=" << labels[i] << ", distance=" << distances[i] << std::endl;
    }
}

void test_different_configurations() {
    std::cout << "\n=== Testing Different HENN Configurations ===" << std::endl;
    
    int d = 32;
    int n = 5000;
    int M = 16;
    
    std::vector<float> database(n * d);
    generate_random_data(database, n, d);
    
    std::vector<float> queries(10 * d);
    generate_random_data(queries, 10, d);
    
    // Test different configurations
    std::vector<std::pair<std::string, std::string>> configs = {
        {"random", "nsw"},
        {"random", "knn"},
        {"budget_aware", "nsw"},
        {"budget_aware", "knn"}
    };
    
    for (const auto& config : configs) {
        std::cout << "\nTesting EPS-Net: " << config.first 
                  << ", Graph: " << config.second << std::endl;
        
        faiss::IndexHENNFlat index(d, M, faiss::METRIC_L2);
        index.set_epsnet_strategy(config.first);
        index.set_pgraph_algorithm(config.second);
        index.henn.ef_construction = 100;
        
        auto start = std::chrono::high_resolution_clock::now();
        index.add(n, database.data());
        auto end = std::chrono::high_resolution_clock::now();
        
        auto build_time = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
        std::cout << "  Build time: " << build_time.count() << " ms" << std::endl;
        std::cout << "  Layers: " << index.henn.layers.size() << std::endl;
        
        // Quick search test
        std::vector<float> distances(10);
        std::vector<faiss::idx_t> labels(10);
        
        start = std::chrono::high_resolution_clock::now();
        index.search(1, queries.data(), 10, distances.data(), labels.data());
        end = std::chrono::high_resolution_clock::now();
        
        auto search_time = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
        std::cout << "  Search time: " << search_time.count() << " μs" << std::endl;
    }
}

void test_comparison_with_flat() {
    std::cout << "\n=== Comparing HENN with Flat Index ===" << std::endl;
    
    int d = 128;
    int n = 10000;
    int nq = 100;
    int k = 10;
    
    std::vector<float> database(n * d);
    std::vector<float> queries(nq * d);
    generate_random_data(database, n, d);
    generate_random_data(queries, nq, d);
    
    // Test flat index
    std::cout << "Testing IndexFlatL2..." << std::endl;
    faiss::IndexFlatL2 flat_index(d);
    
    auto start = std::chrono::high_resolution_clock::now();
    flat_index.add(n, database.data());
    auto end = std::chrono::high_resolution_clock::now();
    auto flat_build = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    
    std::vector<float> flat_distances(nq * k);
    std::vector<faiss::idx_t> flat_labels(nq * k);
    
    start = std::chrono::high_resolution_clock::now();
    flat_index.search(nq, queries.data(), k, flat_distances.data(), flat_labels.data());
    end = std::chrono::high_resolution_clock::now();
    auto flat_search = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
    
    std::cout << "  Build time: " << flat_build.count() << " ms" << std::endl;
    std::cout << "  Search time: " << flat_search.count() << " μs" << std::endl;
    
    // Test HENN index
    std::cout << "Testing IndexHENNFlat..." << std::endl;
    faiss::IndexHENNFlat henn_index(d, 16, faiss::METRIC_L2);
    henn_index.henn.ef_construction = 200;
    
    start = std::chrono::high_resolution_clock::now();
    henn_index.add(n, database.data());
    end = std::chrono::high_resolution_clock::now();
    auto henn_build = std::chrono::duration_cast<std::chrono::milliseconds>(end - start);
    
    std::vector<float> henn_distances(nq * k);
    std::vector<faiss::idx_t> henn_labels(nq * k);
    
    faiss::SearchParametersHENN search_params;
    search_params.ef = 50;
    
    start = std::chrono::high_resolution_clock::now();
    henn_index.search(nq, queries.data(), k, henn_distances.data(), henn_labels.data(), &search_params);
    end = std::chrono::high_resolution_clock::now();
    auto henn_search = std::chrono::duration_cast<std::chrono::microseconds>(end - start);
    
    std::cout << "  Build time: " << henn_build.count() << " ms" << std::endl;
    std::cout << "  Search time: " << henn_search.count() << " μs" << std::endl;
    
    // Calculate recall
    int correct = 0;
    for (int i = 0; i < nq; i++) {
        for (int j = 0; j < k; j++) {
            for (int l = 0; l < k; l++) {
                if (henn_labels[i * k + j] == flat_labels[i * k + l]) {
                    correct++;
                    break;
                }
            }
        }
    }
    
    double recall = (double)correct / (nq * k);
    std::cout << "  Recall@" << k << ": " << recall << std::endl;
    std::cout << "  Speedup: " << (double)flat_search.count() / henn_search.count() << "x" << std::endl;
}

int main() {
    std::cout << "HENN Index Example" << std::endl;
    std::cout << "==================" << std::endl;
    
    try {
        test_henn_flat();
        test_different_configurations();
        test_comparison_with_flat();
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
    
    std::cout << "\nAll tests completed successfully!" << std::endl;
    return 0;
}
